### ISSUE TITLE
Use cosmos db as datastore

### ISSUE THREAD

**Issue by timClyburn** [Azure/azure-mobile-apps#450]
## Describe the bug

I am attempting to use cosmos db as the datastore for datasync which should be supported according to the docs:
[Supported Platforms](https://docs.microsoft.com/en-gb/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#supported-platforms)
![image](https://user-images.githubusercontent.com/28535141/185181119-0d473e3e-5c72-494e-bd87-ab78eb86e95f.png)

I have changed the connection strings to point to the cosmos db and added the required nuget packages. When I POST an item to the endpoint the data gets added to the database correctly but does not sync.
`builder.Services.AddDbContext<AppDbContext>(options => options.UseCosmos(connectionString, "MobileData"));`

## To Reproduce

Follow the [guide](https://docs.microsoft.com/en-gb/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core) to create a new Datasync server.
Add the Microsoft.EntityFrameworkCore.Cosmos nuget package
Configure the cosmos connection by replacing the line
`builder.Services.AddDbContext<AppDbContext>(options => options.UseSqlServer(connectionString));`
with
`builder.Services.AddDbContext<AppDbContext>(options => options.UseCosmos(connectionString, "MobileData"));`
And update the connection string to be a cosmos connection string

You will get the below exception:
`System.InvalidOperationException: 'The entity type 'TodoItem' has property 'Version' configured as a concurrency token, but only a property mapped to '_etag' is supported as a concurrency token. Consider using 'PropertyBuilder.IsETagConcurrency'.'`

Override the OnModelCreating function to set the etag as the version property:
``` c#
protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<TodoItem>()
                .Property(t => t.Version)
                .IsETagConcurrency();
            base.OnModelCreating(modelBuilder);
        }
```
This gives the error:
`System.InvalidOperationException: 'The type of the etag property 'Version' on 'TodoItem' is 'byte[]'. All etag properties must be strings or have a string value converter.'`

I have added `HasConversion()` in multiple different formats to the propertybinding but nothing seems to permit fetching the data.

This gives the below error:
``` text
fail: Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'MobileAppService.Db.AppDbContext'.
      System.FormatException: The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.
         at System.Convert.FromBase64CharPtr(Char* inputPtr, Int32 inputLength)
         at System.Convert.FromBase64String(String s)
         at lambda_method95(Closure , QueryContext , JObject )
         at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
      System.FormatException: The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.
         at System.Convert.FromBase64CharPtr(Char* inputPtr, Int32 inputLength)
         at System.Convert.FromBase64String(String s)
         at lambda_method95(Closure , QueryContext , JObject )
         at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
```
Or this, when I post data (post succeeds and data is in the database):
``` text
fail: Microsoft.EntityFrameworkCore.Update[10000]
      An exception occurred in the database while saving changes for context type 'MobileAppService.Db.AppDbContext'.
      Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the the item with id 'TodoItem|1982aba5a31b4f1bb2215a25d5c8cf62'. See the inner exception for details.
       ---> System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Byte[]'.
         at Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer`1.Snapshot(Object instance)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SidecarValues.SetValue(IProperty property, Object value, Int32 index)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetProperty(IPropertyBase propertyBase, Object value, Boolean isMaterialization, Boolean setModified, Boolean isCascadeDelete, CurrentValueType valueType)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetStoreGeneratedValue(IProperty property, Object value)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.ProcessResponse(ResponseMessage response, IUpdateEntry entry)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.CreateItemOnceAsync(DbContext _, ValueTuple`4 parameters, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.CreateItemOnceAsync(DbContext _, ValueTuple`4 parameters, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosDatabaseWrapper.SaveChangesAsync(IList`1 entries, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosDatabaseWrapper.SaveChangesAsync(IList`1 entries, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.<>c__DisplayClass33_0`2.<<ExecuteAsync>b__0>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
      Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while saving the the item with id 'TodoItem|1982aba5a31b4f1bb2215a25d5c8cf62'. See the inner exception for details.
       ---> System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Byte[]'.
         at Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer`1.Snapshot(Object instance)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SidecarValues.SetValue(IProperty property, Object value, Int32 index)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetProperty(IPropertyBase propertyBase, Object value, Boolean isMaterialization, Boolean setModified, Boolean isCascadeDelete, CurrentValueType valueType)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetStoreGeneratedValue(IProperty property, Object value)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.ProcessResponse(ResponseMessage response, IUpdateEntry entry)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.CreateItemOnceAsync(DbContext _, ValueTuple`4 parameters, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosClientWrapper.CreateItemOnceAsync(DbContext _, ValueTuple`4 parameters, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosDatabaseWrapper.SaveChangesAsync(IList`1 entries, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.EntityFrameworkCore.Cosmos.Storage.Internal.CosmosDatabaseWrapper.SaveChangesAsync(IList`1 entries, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(StateManager stateManager, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.<>c__DisplayClass33_0`2.<<ExecuteAsync>b__0>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
```

## Expected behavior

Post and Get to happen without errors

## What platforms?

* Server (ASP.NET Framework, ASP.NET Core, NodeJS):
  * Library versions?
    *   `<PackageReference Include="Microsoft.AspNetCore.Datasync" Version="5.0.10" />`
    *   `<PackageReference Include="Microsoft.AspNetCore.Datasync.EFCore" Version="5.0.10" />`
    *   `<PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.8" />`
    *  `<PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="6.0.8">`
  * What database are you using? Cosmos
  * Where are you running the server? Local dev machine

* Client POSTMAN

**Comment by adrianhall** [comment]
Thanks - looks like I need to have a specific version of the Entity Framework repository for Cosmos since it doesn't like the `byte[]` etag.  The new provider will convert between the `byte[]` version and the string version.  You may be able to use the Mapped provider as well.

Will work on it.

**Comment by adrianhall** [comment]
To be done:

1) Implement "ETagEntityTableData" to convert between Version and EntityTag
2) Add type checks into EntityTableRepository to ensure that the provided generic type is either EntityTableData or ETagEntityTableData
3) Add tests for version to ensure conversion between the two table types.
4) Add tests for entity type checking
5) Update the docs to reflect Cosmos specific setup (IsETagConcurrency, etc.)

**Comment by adrianhall** [comment]
Added a new `ITableData` implementation called `ETagEntityTableData` which includes the `EntityTag` field (which you don't need to worry about, except it exists).  To implement the Cosmos DB driver:

1. Ensure that your model derives from `ETagEntityTableData`
2. Add an `OnModelCreating(ModelBuilder builder)` as follows:

    ``` csharp
        protected override void OnModelCreating(ModelBuilder builder)
        {
            builder.Entity<TodoItem>()
                .Property(t => t.EntityTag)
                .IsETagConcurrency();
            base.OnModelCreating(builder);
        }
    ```

Repeat the `IsETagConcurrency()` check for each model you are exposing.  You can do other things in there as well, like set partition keys and containers on a per-table basis.  

This will all be released in 5.0.11 - I'll also include a CosmosBasedServer sample in the samples directory, which will get published this afternoon.

**Comment by adrianhall** [comment]
v5.0.11 approved for release to NuGet with fix.
